### PR TITLE
[ECAM] Improve ECAM Memo TO Flaps and TO CONFIG checks

### LINKED ISSUE #1569: [BUG] Identified Possible T.O Config Discrepancy
Mod Version - Master

**Describe the bug**
Through some QA testing I was doing of the general Master and of the XML pull PR via QA yesterday I operated into and out of a number of airports testing as many features as possible.

One test included a takeoff out of LGSK - which has a runway length of just over 1500m - quite short but I wanted to test performance, config etc for short field use as it's not been tested much.

Started Cold and Dark - Configured the MCDU to indicate a calculated 61° flex temp and FLAPS 3 for the short runway departure as I understand would be normal IRL and updated the Vrefs accordingly in the MCDU.

During the short taxi I set up for takeoff and selected TO Config test to check TO Config and for it to show "Norm" on the eicas quick checklist.

I got Master Warning and visual warnings on eicas for takeoff config not being correct. 

I reduced to Flaps 2 - pressed TO Config again and it cleared everything, and after this I returned to Flaps 3 for the departure.

**To Reproduce**

1. Spawn at any runway and configure as if to use a very short runway departure requiring Flaps 3. 
2. Ensure Flaps 3 set in MCDU on perf page and everything else is configured accordingly
3. During taxi, press the TO Config Test button 

**Expected behavior**
Set Flaps 3 in TO Perf page due to very short 1500m runway departure requiring F3. Prior to departure, test TO Config and it should show as Norm without issue. 

F3 is available for this purpose for short runway takeoff IRL I understand, so wanted to specifically test the mod behaviour with these conditions as a QA.

**Actual behavior**
It did not accept a Flap 3 config for very short takeoff even with Flap 3 set in the MCDU TO Perf Page.
When running TO Config Test again reselecting Flap 2 that worked but I needed Flaps 3 - so when I reset that again I got Master Warning again.

I cleared the issue (workaround) by selecting Flap 2 with the config warning displayed, retested TO Config to clear the message as it accepted F2 - then reselected F3 afterwards.

**References**
Found this just now which helps explain what I've done in a million words above! That's how I expected it to behave (I also used LGSK for my QA test flight when I discovered this issue)
https://youtu.be/myBmuhsJf90[https://youtu.be/myBmuhsJf90](https://youtu.be/myBmuhsJf90) 

**Additional context**
N/A

Was this working before/when did the issue start occurring?
Just noticed it today from running different QA tests for the aircraft development 

Is this a problem in the vanilla unmodded game?
No

Discord username (if different from GitHub): 
British Avgeek YT - Carter633 #1436
(I'm on the QA chat on discord and have also asked Danbw9 to take a look to see how it behaves with his knowledge as IRL pilot) give me a shout on there if you need anything. I also recorded the behaviour on my YouTube channel though the video isn't live yet

### PR DESCRIPTION
**Pull request by MisterChocker**
Fixes #1569

## Summary of Changes

- Improved ECAM Memo TO Flaps
- Improved TO CONFIG checks now for physical flaps and slats positions
- Allows for TO flaps config 3

## References

ESLD

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Check various take off configs.
Allowed take off configs are 1 + F, 2 and 3
Check for Flap or Slat not in to config ecam warning and ecam memo blue.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Review by Benjozork — APPROVED** [review]
LGTM

**Comment by marinerhally** [comment]
Category: QA Tester
Name: Discord User Mariner Hally#0715
Date of testing: 13/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3016
Object of test: Improve TO flaps config (Allow flaps 3 T/O).

Testing technique: Spawned in C/D and tested T/O configs.
Steps you've made to test the Pull Request:
1. Spawned in at various airports. 
2. Tested multiple different T/O configs.
3. Tested trying to T/O with wrong flaps config and all correct warnings sounded on each occasion.

Flight notes (If required): Managed to take off Flaps 3 on 1020m :)

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: No problems at all with this PR.

**Comment by tjburni** [comment]
Category: QA Tester
Name: [Z+1]tjburni#0409
Date of testing: Current date 13.1.2021
Version of the sim: 1.12.13
PR Tested: #3016
Object of test:TO Flaps config. Allow Flaps 3 for TO

Testing technique: tested?
Steps you've made to test the Pull Request:
1. spawned at 5 different runways 
2. tested 2 times on each airport with different T/O configs
3. Checked the warning sounds for wrong Flap settings
4.
5.

Flight notes (If required):

Issues (If Any): none
Overall Rating: Good/To be improved
Conclusions of the testing:Works as expected. On a short runway i took off qith flaps3. Thanks for the PR!

**Comment by MisterChocker** [comment]
> Good/To be improved

So is it good then? From the consensus of your report I assume the result should be Good, but you also wrote To be improved.

**Comment by tjburni** [comment]
> > Good/To be improved
> 
> So is it good then? From the consensus of your report I assume the result should be Good, but you also wrote To be improved.

Sorry i forgot that XD

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -32,6 +32,7 @@
 1. [MCDU] Fixed APPR/GO AROUND page order - @beheh (Benedict Etzel)
 1. [OPTIONS] Changed unit option changes only effective after aircraft reload - @MisterChocker (Leon)
 1. [ECAM] Fixed engineering mode overlay displaying on upper ECAM in some cases - @davidwalschots (David Walschots)
+1. [ECAM] Added Config 3 TO config - @MisterChocker (Leon)
```
Comment (MisterChocker) [inline .github/CHANGELOG.md:35]:
```suggestion
1. [ECAM] Improved Ecam Memo TO Flaps and TO Config. Allows for Config 3 TO - @MisterChocker (Leon)
```

**Comment by beheh** [comment]
For QA testers: Note that the flaps value you enter PERF TAKEOFF is (correctly!) ignored for the sake of this PR. There will be a separate ECAM warning coming in future that checks that the value you entered in the MCDU matches your flaps, but that's out of scope for this PR. The only thing the warning does here is that it enforces a minimum and maximum physical flap and slat position on the aircraft, to ensure the aircraft is physically configured for flight.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 2]
```diff
@@ -616,6 +616,20 @@ var A320_Neo_UpperECAM;
                     {
                         name: "CONFIG",
                         messages: [
+                            {
+                                id: "config_slats",
+                                message: "",
+                                level: 3,
+                                flightPhasesInhib: [5, 6, 7, 8, 10], // TODO
```
Comment (beheh) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:623]:
```suggestion
                                flightPhasesInhib: [5, 6, 7, 8], // TODO
```
You can drop the 10 here and below according to my docs, but it's not a big deal.

**Review by beheh — APPROVED** [review]

**Comment by Awemeter** [comment]
Category: QA Tester
Name: Discord User #0000
Date of testing: Current date DD/MM/YYYY
Version of the sim: Version the tester uses
PR Tested: #000
Object of test: (ECAM Page or Flight model or views or anything, what the PR modifies)

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Spawned C&D at katl and loaded simbreif KATL to KIAD
2. Programmed take off flap and trim in perf page then taxied to the runway and took off 
3. Repeated prosses with Flap 1,2,3.
4. Checked ecam memo for Flap/Slat, not in config.
5. Moved flaps in and out of take-off position

Flight notes (If required):


Issues (If Any): If you have your slats out of position it will only give you a warning if you have it in flaps full.
Overall Rating: To be improved
Conclusions of the testing: It allows for config 3 but the ecam memo and to config do not work properly. Great pr keep up the good work

**Comment by beheh** [comment]
Please be specific. When trying to describe an issue encountered, please *exactly* state which configuration the aircraft was in and what you did, otherwise we can only guess what you mean.

> If you have your slats out of position it will only give you a warning if you have it in flaps full.

What does this mean? What exactly did you do? What did flaps full fix?

> It allows for config 3 but the ecam memo and to config do not work properly. Great pr keep up the good work

This is the first time you mention the ECAM memo having an issue. What's the issue?

**Comment by Awemeter** [comment]
When I set up config 1,2,3 in the mcdu it would not give me a warning when I took of with flaps out of config. It only gave me a warning when I put flaps full.

**Comment by MisterChocker** [comment]
@Awemeter 

> For QA testers: Note that the flaps value you enter PERF TAKEOFF is (correctly!) ignored for the sake of this PR. There will be a separate ECAM warning coming in future that checks that the value you entered in the MCDU matches your flaps, but that's out of scope for this PR. The only thing the warning does here is that it enforces a minimum and maximum flap and slat position on the aircraft, to ensure the aircraft is physically configured for flight.

**Comment by cianwhite** [comment]
Category: QA Tester
Name: Cian#3596
Date of testing: 13/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3016 
Object of test: Changes to ECAM TO Flaps and TO Config push button

Testing technique: Description of how you tested the Pull Request

1. Started C+D at EIDW. Started as normal and tried entering "0" and "4" in TO PERF page, which resulted in an error message as expected. Then I entered flaps 1,2 ,3 and they all worked and the adjusted V1, VR, V2 speeds accordingly.
2. Tried the TO config test with each of these flaps settings and only received a "Flaps not in TO config" and "Slats not in TO config" when on Flaps 0 and Flaps 4, setting them to 1,2,3 fixes this as expected.
3. Successfully took off on Flaps 1,2 and 3.

Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: I encountered the TO warning only on Flaps 0 and Flaps 4, which is to be expected. The TO config push button showed the same results, only encountering warnings on Flap 0 and 4. Good job :)

**Comment by NickFarin** [comment]
Category: QA Tester
Name: Discord User: Nick F[Z+8]#5327
Date of testing: Current date 13/01/20
Version of the sim: 1.12.13.0
PR Tested: #3016 
Object of test: ECAM memo and TO CONFIG
Testing technique: Spawned at runway 05 and tested different flaps settings
Steps you've made to test the Pull Request:
1. Spawned at the runway at RPMD
2. Set up the PERF page on the MCDU
3. pushed the TO CONFIG button to check for master warning and ECAM warning is flaps are up or set at 4
4. checked if flaps 1+F, 2, and 3 works fine

Flight notes (If required):

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: The test was normal and flaps 1+F to 3 works fine as well as the ECAM warning if the improper flaps setting is set.

**Comment by BritishAvgeek** [comment]
Category: QA Tester
Name: Discord User: BritishAvgeek#1436
Date of testing: Current date 14/01/21
Version of the sim: 1.12.13.0
PR Tested: #3016
Object of test: ECAM memo and TO CONFIG bug fix

Testing technique: Spawned C&D at EGBB and configured for a Flap 3 departure, took off. Respawned at LGSK and re-checked
Steps you've made to test the Pull Request:
1. Spawned C&D at EGBB. Configured for a dummy flight to EGPH. Set configuration for a Flap 3 departure
2. Selected F3 during taxi, checked TO Config. Accepted the config but other issues presented...
3. Departed to check for the other side of this bug I reported where it then re-activated Master Warning, no issues in the F3 config
4. Respawned at LGSK and re-checked same process.

Flight notes (If required):

Issues (If Any): With Flaps 3 entered into the MCDU, the F3 configuration was accepted by the ECAM TO Config Test and displayed Normal however when I then dropped it to F2 and re-checked, it remained on Normal. I checked 1+F and rechecked and it showed Normal still. When F0 and retested, it triggered the Master Warning.

When I cleared that and only selected F1, it again only showed TO Config Normal for every single flap setting except F4. Again, it accepted F3 but did for everything between that and F0 unfortunately.
During the takeoff roll, F3 presented no issues but interestingly when I selected F1+F and F2 on two subsequent departure attempts in the wrong config (config 3 set) it didn't trigger any Master Warning as it should have done as the config was wrong.

Overall Rating: To be Improved
Conclusions of the testing: Thanks for starting to fix this bug - it was an important one for those shorter runway departures like LGSK!

**Comment by beheh** [comment]
> Edit: I didn't see the bit about how the PR should ignore the MCDU config flap setting... so in that essence it worked correctly... but then I don't quite get the logic behind this PR as you'd want the Master Warning to trigger if you try to take off with F2 when you'd need or have set F3 in the MCDU... this PR makes the aircraft not do that?

The PR has been corrected since it was initially being tested. In a previous version, this PR incorrectly cross-checked the MCDU, which is not accurate for the specific warning that was touched for this PR. As far as I know @MisterChocker has elected to handle the MCDU cross-check in a separate PR, so this PR is only about the warning here.

Reading #1569 I feel like that has been adequately fixed, and there has been no issue filed so far for the explicit MCDU cross-check warning.

**Pull request merged**